### PR TITLE
Fixed multiple JSON-related issues

### LINKED ISSUE #15677: JSONB UPDATE ERROR
### What steps will reproduce the problem?
I am trying to update JSONB data column in the Postgres database using ActiveRecord.
### What is the expected result?
I expect success on update
### What do you get instead?
```JavaScript
{
    "name": "Database Exception",
    "message": "Object of class yii\\db\\JsonExpression could not be converted to string\nFailed to prepare SQL: UPDATE \"crm\".\"deal\" SET \"search_query\"=:qp0, \"updated_at\"=:qp1 WHERE \"id\"=:qp2",
    "code": 4096,
    "type": "yii\\db\\Exception",
    "file": "/home/borodulin/projects/roomberry/vendor/yiisoft/yii2/db/Command.php",
    "line": 264,
    "stack-trace": [
        "#0 /home/borodulin/projects/roomberry/vendor/yiisoft/yii2/db/Command.php(1067): yii\\db\\Command->prepare(false)",
        "#1 /home/borodulin/projects/roomberry/vendor/yiisoft/yii2/db/ActiveRecord.php(247): yii\\db\\Command->execute()",
        "#2 /home/borodulin/projects/roomberry/vendor/yiisoft/yii2/db/BaseActiveRecord.php(801): yii\\db\\ActiveRecord::updateAll(Array, Array)",
        "#3 /home/borodulin/projects/roomberry/vendor/yiisoft/yii2/db/ActiveRecord.php(591): yii\\db\\BaseActiveRecord->updateInternal(NULL)",
        "#4 /home/borodulin/projects/roomberry/vendor/yiisoft/yii2/db/BaseActiveRecord.php(672): yii\\db\\ActiveRecord->update(true, NULL)",
        "#5 /home/borodulin/projects/roomberry/vendor/yiisoft/yii2/rest/UpdateAction.php(48): yii\\db\\BaseActiveRecord->save()",
        "#6 [internal function]: yii\\rest\\UpdateAction->run('47135')",
        "#7 /home/borodulin/projects/roomberry/vendor/yiisoft/yii2/base/Action.php(94): call_user_func_array(Array, Array)",
        "#8 /home/borodulin/projects/roomberry/vendor/yiisoft/yii2/base/Controller.php(157): yii\\base\\Action->runWithParams(Array)",
        "#9 /home/borodulin/projects/roomberry/vendor/yiisoft/yii2/base/Module.php(528): yii\\base\\Controller->runAction('update', Array)",
        "#10 /home/borodulin/projects/roomberry/vendor/yiisoft/yii2/web/Application.php(103): yii\\base\\Module->runAction('crm/deal/update', Array)",
        "#11 /home/borodulin/projects/roomberry/vendor/yiisoft/yii2/base/Application.php(386): yii\\web\\Application->handleRequest(Object(yii\\web\\Request))",
        "#12 /home/borodulin/projects/roomberry/api/web/index.php(18): yii\\base\\Application->run()",
        "#13 {main}"
    ],
    "error-info": null,
    "previous": {
        "name": "PHP Recoverable Error",
        "message": "Object of class yii\\db\\JsonExpression could not be converted to string",
        "code": 4096,
        "type": "yii\\base\\ErrorException",
        "file": "/home/borodulin/projects/roomberry/vendor/yiisoft/yii2/db/Command.php",
        "line": 316,
        "stack-trace": [
            "#0 /home/borodulin/projects/roomberry/vendor/yiisoft/yii2/db/Command.php(260): yii\\db\\Command->bindPendingParams()",
            "#1 /home/borodulin/projects/roomberry/vendor/yiisoft/yii2/db/Command.php(1067): yii\\db\\Command->prepare('???')",
            "#2 /home/borodulin/projects/roomberry/vendor/yiisoft/yii2/db/ActiveRecord.php(247): yii\\db\\Command->execute()",
            "#3 /home/borodulin/projects/roomberry/vendor/yiisoft/yii2/db/BaseActiveRecord.php(801): yii\\db\\ActiveRecord::updateAll('???', '???', '???')",
            "#4 /home/borodulin/projects/roomberry/vendor/yiisoft/yii2/db/ActiveRecord.php(591): yii\\db\\BaseActiveRecord->updateInternal('???')",
            "#5 /home/borodulin/projects/roomberry/vendor/yiisoft/yii2/db/BaseActiveRecord.php(672): yii\\db\\ActiveRecord->update('???', '???')",
            "#6 /home/borodulin/projects/roomberry/vendor/yiisoft/yii2/rest/UpdateAction.php(48): yii\\db\\BaseActiveRecord->save('???', '???')",
            "#7 /home/borodulin/projects/roomberry/vendor/yiisoft/yii2/base/Action.php(94): yii\\rest\\UpdateAction->run('???')",
            "#8 /home/borodulin/projects/roomberry/vendor/yiisoft/yii2/base/Action.php(94): ::call_user_func_array:{/home/borodulin/projects/roomberry/vendor/yiisoft/yii2/base/Action.php:94}('???', '???')",
            "#9 /home/borodulin/projects/roomberry/vendor/yiisoft/yii2/base/Controller.php(157): yii\\base\\Action->runWithParams('???')",
            "#10 /home/borodulin/projects/roomberry/vendor/yiisoft/yii2/base/Module.php(528): yii\\base\\Controller->runAction('???', '???')",
            "#11 /home/borodulin/projects/roomberry/vendor/yiisoft/yii2/web/Application.php(103): yii\\base\\Module->runAction('???', '???')",
            "#12 /home/borodulin/projects/roomberry/vendor/yiisoft/yii2/base/Application.php(386): yii\\web\\Application->handleRequest('???')",
            "#13 /home/borodulin/projects/roomberry/api/web/index.php(18): yii\\base\\Application->run()",
            "#14 {main}"
        ]
    }
}
```

### Additional info

| Q                | A
| ---------------- | ---
| Yii version      | 2.0.14
| PHP version      | 7.0
| Operating system | Ubuntu 16.04 LTS

### LINKED ISSUE #15684: JsonExpression in Command::update() method
### What steps will reproduce the problem?
Update postgres table contains JSON(B) object using  $connection->createCommand()->update () method

### What is the expected result?

Updated record

### What do you get instead?

Database Exception – yii\db\Exception
Object of class yii\db\JsonExpression could not be converted to string
Failed to prepare SQL: UPDATE "object_attributes" SET "attr_value"=:qp2 WHERE "UUID" = :uuid and "attr_name"=:attr_name
↵
Caused by: yii\base\ErrorException
Object of class yii\db\JsonExpression could not be converted to string
in /private/var/www/zzz/www/vendor/yiisoft/yii2/db/Command.php at line 315

1. in /private/var/www/zzz/www/vendor/yiisoft/yii2/db/Command.php at line 263

### LINKED ISSUE #15707: MySQL returns JSON type field as a string
### Additional info

| Q                | A
| ---------------- | ---
| Yii version      | 2.0.14

According to [doc](http://www.yiiframework.com/doc-2.0/guide-db-active-record.html#json-in-mysql-and-postgresql) for mysql and postgre JSON fields should be decoded. But there is [JSON::decode() for PostgreSQL](https://github.com/yiisoft/yii2/tree/2.0.14/framework/db/pgsql/ColumnSchema.php#L90) only, not MySQL.

### PR DESCRIPTION
**Pull request by SilverFire**
This PR fixes multiple problems related to JSON handling in ActiveRecord layer, reported in the following issues:

- [x] Wait for tests to pass
- [x] Fix test for Optimistic lock (not blocks PR from review)

Fixes #15677, fixes #15684, fixes #15707
Replaces #15712

### CODE REVIEW

**Comment by Illuminerdy** [comment]
I have a similar problem, has  it been addressed?

[yii\db\Exception] Object of class yii\db\JsonExpression could not be converted to string
Failed to prepare SQL: UPDATE "widget_layouts" SET "data"=:qp0 WHERE "id"=:qp1

**Comment by SilverFire** [comment]
Yes

**Comment by Illuminerdy** [comment]
We had to redo a lot of code to fix this code change. The JsonExpression should use the toString overload so we don't to call 'getValue' for old code. This caused us a big headache. Our code was using json_encode and decode everywhere.

File: framework/db/QueryBuilder.php [hunk 1]
```diff
@@ -624,15 +624,16 @@ protected function prepareUpdateSets($table, $columns, $params = [])
         $columnSchemas = $tableSchema !== null ? $tableSchema->columns : [];
         $sets = [];
         foreach ($columns as $name => $value) {
+            $columnName = $this->db->quoteColumnName($name);
```
Comment (samdark) [inline framework/db/QueryBuilder.php]: Should be moved down to `$sets[] =`.

**Review by samdark — APPROVED** [review]
Looks good but please apply the optimization I've commented about.

**Comment by SilverFire** [comment]
Will merge today evening

**Pull request merged**